### PR TITLE
Add kubebuilder API definitions

### PR DESCRIPTION
**Pull request by alexander-demichev**
These are API definitions that we can use with controller runtime controllers. They are mostly identical and should be compatible. The main differences are:

- Conditions use structures provided by the upstream metav1 package.
- All conditions types and reasons are now constants, making for us easier to manipulate them.
- Machine inventory `status.PlanStatus` now contains `planState` and only one `checksum` field.
- Some fields include additional validation using kubebuilder tags.

The PR doesn't include `ManagedOSImage` and `ManagedOSVersionChannel` as they are not implemented yet and
`Config` types need some documentation.

Fixes: https://github.com/rancher/elemental-operator/issues/178

### CODE REVIEW

**Review by mjura — APPROVED** [review]
LGTM

**Review by fgiudici — APPROVED** [review]
just left a couple of comments, but looks good here, nice PR!

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,5 @@
 build/
 coverage.out
-.idea/*
\ No newline at end of file
+.idea/*
+hack/tools/bin/
+.vscode/
```
[the remark this answers is not in the document]
Comment (alexander-demichev) [reply .gitignore:5]: I store my debugger config there :) can we keep it?

File: api/v1beta1/types.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	Reboot bool `json:"reboot,omitempty"`
+	// +optional
+	EjectCD bool `json:"eject-cd,omitempty"`
+}
+
+type Registration struct {
+	// +optional
+	URL string `json:"url,omitempty"`
+	// +optional
+	CACert string `json:"ca-cert,omitempty"`
```
[the remark this answers is not in the document]
Comment (alexander-demichev) [reply api/v1beta1/types.go:50]: Ok, I can bring them back

[the remark this answers is not in the document]
Comment (alexander-demichev) [reply api/v1beta1/types.go:50]: I can't find where mapstructure fields are used. How having yaml tags affects the behavior?

[the remark this answers is not in the document]
Comment (fgiudici) [reply api/v1beta1/types.go:50]: I think it is ok to add them when we will switch from pkg/config/config.go to here from the elemental register cmd.

[the remark this answers is not in the document]
Comment (fgiudici) [reply api/v1beta1/types.go:50]: I would merge this PR 😄

**Pull request merged**